<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подключение-к-базе-данных" data-toc-modified-id="Подключение-к-базе-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подключение к базе данных</a></span></li><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обзор данных</a></span><ul class="toc-item"><li><span><a href="#Таблица-books" data-toc-modified-id="Таблица-books-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Таблица books</a></span></li><li><span><a href="#Таблица-authors" data-toc-modified-id="Таблица-authors-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Таблица authors</a></span></li><li><span><a href="#Таблица-publishers" data-toc-modified-id="Таблица-publishers-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Таблица publishers</a></span></li><li><span><a href="#Таблица-ratings" data-toc-modified-id="Таблица-ratings-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Таблица ratings</a></span></li><li><span><a href="#Таблица-reviews" data-toc-modified-id="Таблица-reviews-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Таблица reviews</a></span></li></ul></li><li><span><a href="#Запросы-к-базе-данных" data-toc-modified-id="Запросы-к-базе-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Запросы к базе данных</a></span><ul class="toc-item"><li><span><a href="#Посчитайте,-сколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Посчитайте,-сколько-книг-вышло-после-1-января-2000-года-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Посчитайте, сколько книг вышло после 1 января 2000 года</a></span></li><li><span><a href="#Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку" data-toc-modified-id="Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Для каждой книги посчитайте количество обзоров и среднюю оценку</a></span></li><li><span><a href="#Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры" data-toc-modified-id="Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры</a></span></li><li><span><a href="#Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками" data-toc-modified-id="Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками</a></span></li><li><span><a href="#Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок" data-toc-modified-id="Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок</a></span></li></ul></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Выводы</a></span></li></ul></div>

# Описание проекта

Компания решила быть на волне и купила крупный сервис для чтения книг по подписке. 

**Задача работы** — проанализировать базу данных, чтобф сформулировать ценностное предложение для нового продукта.

**База данных** содержит информацию о книгах, издательствах, авторах, а также пользовательские обзоры книг. 

**Описание данных**

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора


## Подключение к базе данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Обзор данных

### Таблица books

In [2]:
query = ''' SELECT *
            FROM books
        '''
ratings = pd.io.sql.read_sql(query, con = engine) 

print(ratings.info())
ratings.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
None


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

### Таблица authors

In [3]:
query = ''' SELECT *
            FROM authors
        '''
ratings = pd.io.sql.read_sql(query, con = engine) 

print(ratings.info())
ratings.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
None


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

### Таблица publishers

In [4]:
query = ''' SELECT *
            FROM publishers
        '''
ratings = pd.io.sql.read_sql(query, con = engine) 

print(ratings.info())
ratings.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
None


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

### Таблица ratings

In [5]:
query = ''' SELECT *
            FROM ratings
        '''
ratings = pd.io.sql.read_sql(query, con = engine) 

print(ratings.info())
ratings.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
None


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

### Таблица reviews

In [6]:
query = ''' SELECT *
            FROM reviews
        '''
ratings = pd.io.sql.read_sql(query, con = engine) 

print(ratings.info())
ratings.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
None


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

## Запросы к базе данных

### Посчитайте, сколько книг вышло после 1 января 2000 года

In [7]:
query = ''' SELECT COUNT(book_id)
            FROM books
            WHERE publication_date >= '2000-01-01'
        '''
pd.io.sql.read_sql(query, con=engine)    

count
0    821

### Для каждой книги посчитайте количество обзоров и среднюю оценку

In [8]:
query = '''   SELECT b.book_id, b.title, AVG(r.rating), COUNT(DISTINCT(re.review_id))
              FROM books AS b LEFT JOIN ratings AS r ON r.book_id = b.book_id 
              LEFT JOIN reviews AS re ON re.book_id = b.book_id
              GROUP BY b.book_id
              ORDER BY COUNT(DISTINCT(re.review_id)) DESC;
        '''
pd.io.sql.read_sql(query, con=engine)    

book_id                                              title       avg  \
0        948                            Twilight (Twilight  #1)  3.662500   
1        963                                Water for Elephants  3.977273   
2        734                                   The Glass Castle  4.206897   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...  4.414634   
4        695  The Curious Incident of the Dog in the Night-Time  4.081081   
..       ...                                                ...       ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel  3.666667   
996      808                            The Natural Way to Draw  3.000000   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites  5.000000   
998      221                          Essential Tales and Poems  4.000000   
999      191  Disney's Beauty and the Beast (A Little Golden...  4.000000   

     count  
0        7  
1        6  
2        6  
3        6  
4        6  
..     ...  
995      0  
996      0  
997      0  
998      0  
999      0  

[1000 rows x 4 columns]

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [9]:
query = ''' SELECT p.publisher, COUNT(b.book_id)
            FROM books AS b JOIN publishers AS p ON b.publisher_id = p.publisher_id
            WHERE num_pages > 50
            GROUP BY p.publisher
            ORDER BY COUNT(b.book_id) DESC
            LIMIT 1
        '''
pd.io.sql.read_sql(query, con=engine)      

publisher  count
0  Penguin Books     42

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [13]:
query = ''' SELECT author, AVG(rating)
            FROM books AS b JOIN ratings AS r ON b.book_id = r.book_id
            JOIN authors AS a ON b.author_id = a.author_id   
            WHERE b.book_id in (SELECT b.book_id
                                FROM books AS b JOIN ratings AS r ON b.book_id = r.book_id               
                                GROUP BY b.book_id
                                HAVING COUNT(rating) >= 50)
            GROUP BY author
            ORDER BY AVG(rating) DESC
            LIMIT 1
        '''
pd.io.sql.read_sql(query, con=engine)   

author       avg
0  J.K. Rowling/Mary GrandPré  4.287097

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [11]:
query = ''' SELECT AVG(s.count)
            FROM (SELECT username, count(review_id) AS count
                  FROM reviews
                  WHERE username in (SELECT username
                                    FROM ratings
                                    GROUP BY username
                                    HAVING COUNT(rating) > 50)
                  GROUP BY username) AS s
        '''
pd.io.sql.read_sql(query, con=engine)    

avg
0  24.333333

## Выводы

В нашем распоряжении было 5 датасетов : **books, authors, publishers, ratings, reviews,** проанализировав их было выявлено:

- **821 книга вышла после 1 января 2000 года**;
- **для каждой книги было посчитано количество обзоров и средняя оценка**;
- **определено издательство, которое выпустило наибольшее число книг толще 50 страниц - издательство Penguin Books, количество книг - 42**;
- **определен автор с самой высокой средней оценкой книг, учтены только книги с 50 и более оценками - J.K. Rowling/Mary GrandPré с средней оценкой 4.287097**;
- **посчитано среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24,33**.